### PR TITLE
MM-46434: Fix permalink bug by replacing the IPs

### PR DESCRIPTION
**Pull request by agarciamontoro**
#### Summary
Opening this in-progress PR right before leaving for holidays to let you take a look while I'm out :)

There's still some code-cleaning to do, some tests to write and some docs to add, but I wanted you to validate the idea in the meantime.

#### Problem

In short, the problem is that when we use a DB dump with posts that contain permalinks, they necessarily point to the IP of the Mattermost server that was used when generating that data. In this case, we'll see posts that look like:

> This is a post that contains a permalink to http://<old_ip>:8065/<team_name>/pl/<post_id>

If we use the dump of this database in a future load-test stressing another server, then resolving that link will fail, since the old server is most probably dead. Then what we want to do is to change the message to something like the following:

> This is a post that contains a permalink to http://<**new**_ip>:8065/<team_name>/pl/<post_id>

#### Solution

For solving this problem, the changes in this PR modify the `comparison` command logic so that instead of loading the DB dump with:

```sh
zcat db_dump.sql.gz | psql -h host -u user dbname
```

it uses `sed` right before feeding the output into `psql`, as in:

```sh
zcat db_dump.sql.gz | sed 's/old_ip/new_ip/g' | psql -h host -u user dbname
```

How does the code build that sed command? Well, we basically need two pieces of data:
- The new IP: this is already in memory, since it's the IP of the deployed app server.
- The old IP: this has to be provided by the user somehow. The proposed solution adds a new `PermalinkIPsToReplace` setting, which is a slice of strings with the IPs that need to be changed (we need a slice because the permalinks may point to the IP of any server in an HA deployment)

With that and a bit of regex magic (which I hope is clear in the comments I added to the code), the solution works as expected. I've tested this running a comparison with both MySQL and PostgreSQL and the IPs are successfully replaced, and I see no errors related to this.

As an aside, I'm thinking that now that we have the DB dump, the S3 bucket, *and* the slice of old IPs, we could have some kind of bundle so that people only need to provide a single file (maybe a tarball with the dump and a config file with the settings needed). In any case, that's something we can do after we merge this PR.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-46434

### CODE REVIEW

File: defaults/validate.go [hunk 1]
```diff
@@ -32,6 +38,14 @@ func Validate(value interface{}) error {
 		}
 	}
 
+	// For non-struct values, we cannot do much, as there's no associated tags
+	// to lookup to decide how to validate, so we have to assume they're valid.
+	if t.Kind() != reflect.Struct {
+		return nil
+	}
```
Comment (agarciamontoro) [inline defaults/validate.go:45]: This change is not directly related to the PR, but we need it to be able to validate slices of basic types, as the slice of strings (`PermalinkIPsToReplace`) this PR adds. It's actually the first time we add a slice of basic types in a config, and that's why this never failed.
The other comment in this file was added because it took me a lot of time to understand that piece of code :nerd_face:

File: defaults/validate_test.go [hunk 2]
```diff
@@ -181,3 +181,37 @@ type testConfig struct {
...
+		S3BucketDumpURI: "s3://test.bucket/subdir",
+		NumUsers:        100,
+		Duration:        "1h",
+	}
+
+	t.Run("valid config", func(t *testing.T) {
+		err := Validate(config)
+		require.NoError(t, err)
+	})
+}
```
Comment (agarciamontoro) [inline defaults/validate_test.go:217]: I'll expand these tests, but I wanted something to check that the changes in validate were actually working with the new `PermalinkIPsToReplace`.

File: comparison/config.go [hunk 3]
```diff
@@ -37,6 +37,12 @@ type LoadTestConfig struct {
 	// This can also point to a local file if prefixed with "file://".
 	// In such case, the dump file will be uploaded to the app servers.
 	DBDumpURL string
+	// An optional list of IPs present in the posts from the DB dump
```
Comment (agnivade) [inline comparison/config.go:40]: How will the user get the list of IPs? Are they expected to go through the DB dump before and find out what IPs are there?

Comment (agarciamontoro) [reply comparison/config.go:40]: My plan is that we (as in the server platform team) bundle all this information together, so devs only have to get a file that contains all this info and use some mechanism to feed that file and let the load-test take care of everything. For what it's worth, I've been getting these IPs by grepping the dump, so we can automate that as soon as the dump is generated.

Comment (agnivade) [reply comparison/config.go:40]: Any reason we can't do this as an internal step? We can grep the dump ourselves, get the IPs, and then replace them. We should try to avoid any manual steps as much as possible.

**Comment by isacikgoz** [comment]
@agarciamontoro would it be possible to use a proxy for that? We can use static host name for these files and on the deployment the proxy can handle which IP it should access to retrieve the files?

**Comment by agarciamontoro** [comment]
@isacikgoz Can you clarify how you envision the usage of the proxy? I understand your idea is to have a site url that's independent of the deployment itself, so that permalinks always point to it. Is that right? But how would we deploy the proxy in that case?

**Comment by isacikgoz** [comment]
> @isacikgoz Can you clarify how you envision the usage of the proxy? 

TBH, haven't put much though on but I was thinking of that we can include additional resource to our terraform file and that proxy would redirects requests to the new URL. In terraform we can configure the proxy once we have the actual URLs.

**Comment by agarciamontoro** [comment]
@isacikgoz The thing is that we would need to keep the site URL from the test with which we generated the dump (so that the permalinks written to the dump point to that proxy) to all the future ones (so that [the permalink detection logic](https://github.com/mattermost/mattermost-server/blob/master/app/post_metadata.go#L565-L644) see the same site URL in the link and identifies it as a permalink, fetching the post internally instead of trying to make a request to what would be understood as an external URL).

I first though that maybe we could solve this doing some VPC magic in AWS with a private address for that potential proxy, but the loadtest agents need to know the site URL as well to do all their actions, so we need something public. We could have a permanent service that somehow redirects to each loadtest instance, but that seems overkill, and I don't think that's what you had in mind.

All in all, I still don't see how a solution other than find-and-replacing in the DB dump could work. But I may be a bit shortsighted with this. If you think this is possible in any other way please let me know.

File: comparison/config.go (hunk 3 above)
Comment (agarciamontoro) [reply comparison/config.go:40]: Do you mean grepping the IPs during the load of the DB dump? I thought about that too, but it would be a bit redundant, since instead of doing it once for each DB dump (when generating it), we'd end up doing it each time we run a test that uses such dump. The time spent in grepping the IPs is not trivial either (I don't remember the specifics, but at least a minute if I'm not mistaken), so I'd rather get all this information when generating the dump and bundling it for the users, so it's transparent for them.
What we need is an automated process to generate the DB dumps, but this PR is the very first step to consume that data, assuming it's already created.
I'm open to change course here too, of course, so please feel free to push back on any of this.

Comment (agnivade) [reply comparison/config.go:40]: >  but this PR is the very first step to consume that data, assuming it's already created.

Ah okay, got you now. Yep, looking forward to the next step. :)

**Review by agnivade — APPROVED** [review]

**Comment by isacikgoz** [comment]
@agarciamontoro thanks for the clarification, let's not complicate things more than necessary then. This looks good to me.

**Comment by agarciamontoro** [comment]
Asking for a re-review, since I added some changes (not super critical, but good to take a look)

**Review by agnivade — APPROVED** [review]
Did not go in-depth. Don't see anything wrong at a quick glance

**Comment by agarciamontoro** [comment]
Thanks :)

**Review by isacikgoz — APPROVED** [review]
lgtm 👍
